src/controlCenter/widgets/menubar/menubar.vala

```diff
@@ -0,0 +1,163 @@
+using GLib;
+
+namespace SwayNotificationCenter.Widgets {
+
+    public struct ConfigObject {
+        string ? name;
+        string ? type;
+        string ? label;
+        string ? position;
+        Action[] actions;
+        bool hidden;
+    }
+
+    public struct Action {
+        string ? label;
+        string ? command;
+    }
+
+    public class Menubar : BaseWidget {
+        public override string widget_name {
+            get {
+                return "menubar";
+            }
+        }
+
+        Gtk.Box menus_container;
+        Gtk.Box topbar_container;
+        List<Gtk.Box> menus;
+
+        ConfigObject[] menu_objects;
+
+        public Menubar (string suffix, SwayncDaemon swaync_daemon, NotiDaemon noti_daemon) {
+            base (suffix, swaync_daemon, noti_daemon);
+
+            Json.Object ? config = get_config (this);
+            if (config != null) {
+                parse_config_objects (config);
+            }
+
+            menus_container = new Gtk.Box (Gtk.Orientation.VERTICAL, 0);
+
+            topbar_container = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 0);
+            topbar_container.get_style_context ().add_class ("topbar");
+
+            menus_container.add (topbar_container);
+
+            foreach (var obj in menu_objects) {
+                add_menu (obj);
+            }
+
+            pack_start (menus_container, true, true, 0);
+            show_all ();
+
+            menus.foreach (m => m.hide ());
+        }
+
+        void add_menu (ConfigObject obj) {
+            if (obj.type == "buttons") {
+                Gtk.Box container = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 0);
+                container.get_style_context ().add_class (obj.name);
+
+                foreach (Action a in obj.actions) {
+                    Gtk.Button b = new Gtk.Button.with_label (a.label);
+
+                    b.clicked.connect (() => {
+                        execute_command (a.command);
+                    });
+
+                    container.add (b);
+                }
+                if (obj.position == "left") {
+                    topbar_container.pack_start (container, false, false, 0);
+                } else if (obj.position == "right") {
+                    topbar_container.pack_end (container, false, false, 0);
+                } else {
+                    debug ("Invalid position for menu item in config");
+                }
+            } else if (obj.type == "menu") {
+
+                Gtk.Button show_button = new Gtk.Button.with_label (obj.label);
+
+                Gtk.Box menu = new Gtk.Box (Gtk.Orientation.VERTICAL, 0);
+                menu.get_style_context ().add_class (obj.name);
+                menus.append (menu);
+                obj.hidden = true;
+
+                show_button.clicked.connect (() => {
+                    if (obj.hidden) {
+                        menus.foreach (m => m.hide ());
```

Comment:
Lets add a nullable menu box into each `ConfigObject` and remove the `menus` list. There's a bug when you're trying to open multiple menus while one's still open which gets solved by this:

```diff
diff --git a/src/controlCenter/widgets/menubar/menubar.vala b/src/controlCenter/widgets/menubar/menubar.vala
index c2b6511..c9e56d8 100644
--- a/src/controlCenter/widgets/menubar/menubar.vala
+++ b/src/controlCenter/widgets/menubar/menubar.vala
@@ -8,7 +8,7 @@ namespace SwayNotificationCenter.Widgets {
         string ? label;
         string ? position;
         Action[] actions;
-        bool hidden;
+        Gtk.Box? menu;
     }
 
     public struct Action {
@@ -25,9 +25,8 @@ namespace SwayNotificationCenter.Widgets {
 
         Gtk.Box menus_container;
         Gtk.Box topbar_container;
-        List<Gtk.Box> menus;
 
-        ConfigObject[] menu_objects;
+        List<ConfigObject?> menu_objects;
 
         public Menubar (string suffix, SwayncDaemon swaync_daemon, NotiDaemon noti_daemon) {
             base (suffix, swaync_daemon, noti_daemon);
@@ -44,17 +43,20 @@ namespace SwayNotificationCenter.Widgets {
 
             menus_container.add (topbar_container);
 
-            foreach (var obj in menu_objects) {
-                add_menu (obj);
+            for (int i = 0; i < menu_objects.length (); i++) {
+                unowned ConfigObject? obj = menu_objects.nth_data (i);
+                add_menu (ref obj);
             }
 
             pack_start (menus_container, true, true, 0);
             show_all ();
 
-            menus.foreach (m => m.hide ());
+            foreach (var obj in menu_objects) {
+                obj.menu?.hide ();
+            }
         }
 
-        void add_menu (ConfigObject obj) {
+        void add_menu (ref unowned ConfigObject? obj) {
             if (obj.type == "buttons") {
                 Gtk.Box container = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 0);
                 container.get_style_context ().add_class (obj.name);
@@ -81,18 +83,14 @@ namespace SwayNotificationCenter.Widgets {
 
                 Gtk.Box menu = new Gtk.Box (Gtk.Orientation.VERTICAL, 0);
                 menu.get_style_context ().add_class (obj.name);
-                menus.append (menu);
-                obj.hidden = true;
+                obj.menu = menu;
 
                 show_button.clicked.connect (() => {
-                    if (obj.hidden) {
-                        menus.foreach (m => m.hide ());
-                        menu.show ();
-                        obj.hidden = !obj.hidden;
-                    } else {
-                        menu.hide ();
-                        obj.hidden = !obj.hidden;
+                    bool visible = !menu.visible;
+                    foreach (var o in menu_objects) {
+                        o.menu?.set_visible (false);
                     }
+                    menu.set_visible (visible);
                 });
 
                 foreach (var a in obj.actions) {
@@ -120,7 +118,7 @@ namespace SwayNotificationCenter.Widgets {
         protected void parse_config_objects (Json.Object config) {
             var elements = config.get_members ();
 
-            menu_objects = new ConfigObject[elements.length ()];
+            menu_objects = new List<ConfigObject?> ();
             for (int i = 0; i < elements.length (); i++) {
                 string e = elements.nth_data (i);
                 Json.Object ? obj = config.get_object_member (e);
@@ -148,14 +146,14 @@ namespace SwayNotificationCenter.Widgets {
                     }
 
                     Action[] actions_list = parse_actions (actions);
-                    menu_objects[i] = ConfigObject () {
+                    menu_objects.append (ConfigObject () {
                         name = name,
                         type = type,
                         label = label,
                         position = pos,
                         actions = actions_list,
-                        hidden = true
-                    };
+                        menu = null,
+                    });
                 }
             }
         }
```

I also changed the array into a GLib List due to some indexes not being filled if that specific menu doesn't get added (due to config error)